lib/puppet-lint/plugins/check_unsafe_interpolations.rb

```diff
@@ -0,0 +1,46 @@
+PuppetLint.new_check(:check_unsafe_interpolations) do
+  COMMANDS = Set['command', 'onlyif', 'unless']
+  def check
+    exec = false
+
+    # Look for exec blocks
+    tokens.select { |token| check_exec?(token) }.each do |token|
+      exec = true
+    end
```

Comment:
I wonder if this could be simplified slightly?

puppet-lint provides a few pre computed indexes that we can work with. One of those is called "resource_indexes".

Maybe you could do something like this to get your exec resources:

```
    exec_resources = resource_indexes.map do |resource|
      resource if resource[:type].value == 'exec'
    end
```

It could even be extended to this:

```
    exec_resources = resource_indexes.map do |resource|
      resource if resource[:type].value == 'exec' && !(COMMANDS & resource[:param_tokens].map(&:value)).empty?
    end
```

You'd potentially end up with an array of resources that can be analysed.